### PR TITLE
Issue-57: root to own app data

### PR DESCRIPTION
**Pull request by marcomc**
added a fix_app_permissions.sh.twig script/template to allow application user ownership and permissions to specific files and directory of the application while the rest of app files are owned by `root`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Harness Base PHP
+
+## Testing
+
+### Static mode
+
+1. run `./build`
+2. Clone https://github.com/inviqa/<framework>-sample in a test directory
```
Comment (kierenevans) [inline README.md]: These are private repos so only relevant to Inviqa employees.

Is there a better way to test this for a member of the public?

Wonder if this works, with `test/` being in `.gitignore` to allow for this:
```
./build
cd test
ws create drupal8-test inviqa/drupal8 --no-install
cp -pR ../dist/harness-drupal8/ drupal8-test/.my127ws/
cd drupal8-test/
ws install
```

File: src/_base/harness/config/confd.yml [hunk 2]
```diff
@@ -16,6 +16,7 @@ confd('harness:/'):
   - { src: docker/image/nginx/root/etc/ssl/certs/app.crt }
   - { src: docker/image/nginx/root/etc/ssl/private/app.key }
   - { src: docker/image/php-fpm/Dockerfile }
+  - { src: docker/image/php-fpm/root/fix_app_permissions.sh}
```
Comment (kierenevans) [inline src/_base/harness/config/confd.yml]:
```suggestion
  - { src: docker/image/php-fpm/root/fix_app_permissions.sh }
```

File: src/drupal8/harness/config/confd.yml [hunk 3]
```diff
@@ -17,6 +17,7 @@ confd('harness:/'):
   - { src: docker/image/nginx/root/etc/ssl/certs/app.crt }
   - { src: docker/image/nginx/root/etc/ssl/private/app.key }
   - { src: docker/image/php-fpm/Dockerfile }
+  - { src: docker/image/php-fpm/root/fix_app_permissions.sh}
```
Comment (kierenevans) [inline src/drupal8/harness/config/confd.yml]:
```suggestion
  - { src: docker/image/php-fpm/root/fix_app_permissions.sh }
```

File: src/magento1/harness.yml [hunk 4]
```diff
@@ -13,6 +13,14 @@ harness('inviqa/magento1'):
 ---
 attributes:
   app:
+    owner: www-data
+    group: www-data
+    dirs_and_files:
+      - '/app/public/app/etc/local.xml'
```
Comment (kierenevans) [inline src/magento1/harness.yml]: Bit strange to have this writeable by magento - https://github.com/continuouspipe/dockerfiles/blob/1e63f8532ec7a6adab4d474987f70540fd2a1271/magento1/usr/local/share/magento1/magento_functions.sh#L16 isn't doing that so perhaps remove this line.
```suggestion
```

File: src/magento1/harness/config/confd.yml [hunk 5]
```diff
@@ -18,6 +18,7 @@ confd('harness:/'):
   - { src: docker/image/nginx/root/etc/ssl/certs/app.crt }
   - { src: docker/image/nginx/root/etc/ssl/private/app.key }
   - { src: docker/image/php-fpm/Dockerfile }
+  - { src: docker/image/php-fpm/root/fix_app_permissions.sh}
```
Comment (kierenevans) [inline src/magento1/harness/config/confd.yml]:
```suggestion
  - { src: docker/image/php-fpm/root/fix_app_permissions.sh }
```

File: src/magento2/harness/config/confd.yml [hunk 6]
```diff
@@ -17,6 +17,7 @@ confd('harness:/'):
   - { src: docker/image/nginx/root/etc/ssl/certs/app.crt }
   - { src: docker/image/nginx/root/etc/ssl/private/app.key }
   - { src: docker/image/php-fpm/Dockerfile }
+  - { src: docker/image/php-fpm/root/fix_app_permissions.sh}
```
Comment (kierenevans) [inline src/magento2/harness/config/confd.yml]:
```suggestion
  - { src: docker/image/php-fpm/root/fix_app_permissions.sh }
```

File: src/spryker/harness/config/confd.yml [hunk 7]
```diff
@@ -16,6 +16,7 @@ confd('harness:/'):
   - { src: docker/image/nginx/root/etc/ssl/certs/app.crt }
   - { src: docker/image/nginx/root/etc/ssl/private/app.key }
   - { src: docker/image/php-fpm/Dockerfile }
+  - { src: docker/image/php-fpm/root/fix_app_permissions.sh}
```
Comment (kierenevans) [inline src/spryker/harness/config/confd.yml]:
```suggestion
  - { src: docker/image/php-fpm/root/fix_app_permissions.sh }
```

File: src/drupal8/harness.yml [hunk 8]
```diff
@@ -9,6 +9,10 @@ harness('inviqa/drupal8'):
 ---
 attributes:
   app:
+    owner: www-data
+    group: www-data
+    dirs_and_files:
+      - '/app/config'
```
Comment (kierenevans) [inline src/drupal8/harness.yml:15]: Also need these: https://github.com/continuouspipe/dockerfiles/blob/3fb53b51bd3b04dbc15855e9f6601fa7fa3cba76/drupal/usr/local/share/drupal/drupal_functions.sh#L39-L40

**Review by kierenevans — CHANGES_REQUESTED** [review]
Made some suggestions :)

File: README.md (hunk 1 above)
Comment (marcomc) [reply README.md]: althought theyt in the dist/application/skeleton
the privisioning reports:
```
fixed permissions for /app/config
/docroot/modules does not exist
/docroot/profiles does not exist
/docroot/sites/default/files does not exist
```

Comment (kierenevans) [reply README.md]: I think we should make the directories if they don't exist (same for all frameworks).

Comment (marcomc) [reply README.md]: I have updated the script to create the missing files and dirs.
I have updated the instructions to create tests environment without using private company repos

File: README.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+1. run `./build`
+2. create and cd a tests directory
+```bash
+mkdir tests
+cd tests
+ws create <framework>-test inviqa/<framework> --no-install
+```
+3. Copy the built version of harness-base-php's ``./dist/<framework>`` dir to `<framework>-test/.my127ws` directory
+```bash
+cp -pR ../dist/harness-<framework>/ <framework>/.my127ws/
```
Comment (kierenevans) [inline README.md]:
```suggestion
cp -pR ../dist/harness-<framework>/ <framework>-test/.my127ws/
```

File: src/drupal8/.ci/sample-static/workspace.yml [hunk 10]
```diff
@@ -2,6 +2,7 @@ workspace('ci-drupal8-sample'):
   description: generated local workspace for ci-drupal8-sample.
   harness: inviqa/drupal8
 
+key('default'): ebbff2496560c9ada246d3783065998cd3799ccabd59c80b7257725218b01095
```
Comment (kierenevans) [inline src/drupal8/.ci/sample-static/workspace.yml]:
```suggestion
```

File: src/drupal8/harness.yml [hunk 11]
```diff
@@ -9,6 +9,14 @@ harness('inviqa/drupal8'):
 ---
 attributes:
   app:
+    owner: www-data
+    group: www-data
+    dirs:
+      - '/app/config'
+      - '/docroot/modules'
```
Comment (kierenevans) [inline src/drupal8/harness.yml]:
```suggestion
      - '/app/docroot/modules'
```

File: src/drupal8/harness.yml [hunk 12]
```diff
@@ -9,6 +9,14 @@ harness('inviqa/drupal8'):
...
+      - '/docroot/profiles'
```
Comment (kierenevans) [inline src/drupal8/harness.yml]:
```suggestion
      - '/app/docroot/profiles'
```

File: src/drupal8/harness.yml [hunk 13]
```diff
@@ -9,6 +9,14 @@ harness('inviqa/drupal8'):
...
+      - '/docroot/sites/default/files'
```
Comment (kierenevans) [inline src/drupal8/harness.yml]:
```suggestion
      - '/app/docroot/sites/default/files'
```

File: README.md [hunk 14]
```diff
@@ -0,0 +1,29 @@
+# Harness Base PHP
+
+## Testing
+
+### Static mode
+
+1. run `./build`
+2. create and cd a tests directory
```
Comment (kierenevans) [inline README.md]:
```suggestion
2. create and change to a tests directory
```

**Comment by kierenevans** [comment]
Wordpress's static build in Jenkins is complaining that:
```
■ > docker build -t blog:develop-php-fpm .
Sending build context to Docker daemon  16.38kB
Step 1/7 : FROM blog:develop-console as console
 ---> 07d3fa7b785c
Step 2/7 : FROM my127/wordpress:7.2-fpm-stretch
 ---> e945a53da66c
Step 3/7 : WORKDIR /app
 ---> Running in f200c9c69bbd
Removing intermediate container f200c9c69bbd
 ---> efaa056caaf7
Step 4/7 : COPY root  /
 ---> d609b574bcd7
Step 5/7 : COPY --from=console --chown=root:root /app /app
 ---> dc19151f44a9
Step 6/7 : RUN bash /fix_app_permissions.sh
 ---> Running in 88fd05f3aa53
bash: /fix_app_permissions.sh: No such file or directory
```

I think we need to have an entry present for all framework's confd.yml to avoid that.

**Comment by kierenevans** [comment]
We should probably get some defaults going in `src/_base/harness/attributes/common.yml`, e.g.:
```
 app:
    owner: www-data
    group: www-data
    dirs: []
    files: []
```

I guess these could also be more obvious, e.g. app.web_writable_dirs and app.web_writable_files ?

**Review by kierenevans — CHANGES_REQUESTED** [review]
Almost there - just the wordpress static build's failure and some more suggestions.

**Comment by marcomc** [comment]
> Wordpress's static build in Jenkins is complaining that:
> 
> ```
> ■ > docker build -t blog:develop-php-fpm .
> Sending build context to Docker daemon  16.38kB
> Step 1/7 : FROM blog:develop-console as console
>  ---> 07d3fa7b785c
> Step 2/7 : FROM my127/wordpress:7.2-fpm-stretch
>  ---> e945a53da66c
> Step 3/7 : WORKDIR /app
>  ---> Running in f200c9c69bbd
> Removing intermediate container f200c9c69bbd
>  ---> efaa056caaf7
> Step 4/7 : COPY root  /
>  ---> d609b574bcd7
> Step 5/7 : COPY --from=console --chown=root:root /app /app
>  ---> dc19151f44a9
> Step 6/7 : RUN bash /fix_app_permissions.sh
>  ---> Running in 88fd05f3aa53
> bash: /fix_app_permissions.sh: No such file or directory
> ```
> 
> I think we need to have an entry present for all framework's confd.yml to avoid that.

done

**Comment by marcomc** [comment]
> We should probably get some defaults going in `src/_base/harness/attributes/common.yml`, e.g.:
> 
> ```
>  app:
>     owner: www-data
>     group: www-data
>     dirs: []
>     files: []
> ```
> 
> I guess these could also be more obvious, e.g. app.web_writable_dirs and app.web_writable_files ?

renamed all the attributes to 
```
web_owner
web_group
web_writable_dirs
web_writable_files
```

**Comment by marcomc** [comment]
> Made some suggestions :)

good suggestions. I have applied them all. Thank you.

**Review by kierenevans — APPROVED** [review]
Great stuff! :tada:

**Pull request merged**